library/psa_crypto.c

```diff
@@ -3882,168 +3425,58 @@ psa_status_t psa_cipher_generate_iv( psa_cipher_operation_t *operation,
                                      size_t iv_size,
                                      size_t *iv_length )
 {
-    psa_status_t status;
-    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-    if( operation->iv_set || ! operation->iv_required )
-    {
-        return( PSA_ERROR_BAD_STATE );
-    }
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
 
-    if( operation->mbedtls_in_use == 0 )
+    if( operation->id == 0 )
     {
-        status = psa_driver_wrapper_cipher_generate_iv( &operation->ctx.driver,
-                                                        iv,
-                                                        iv_size,
-                                                        iv_length );
-        goto exit;
+        return( PSA_ERROR_BAD_STATE );
     }
 
-    if( iv_size < operation->iv_size )
-    {
-        status = PSA_ERROR_BUFFER_TOO_SMALL;
-        goto exit;
-    }
-    ret = mbedtls_psa_get_random( MBEDTLS_PSA_RANDOM_STATE,
-                                  iv, operation->iv_size );
-    if( ret != 0 )
+    if( operation->iv_set || ! operation->iv_required )
     {
-        status = mbedtls_to_psa_error( ret );
-        goto exit;
+        return( PSA_ERROR_BAD_STATE );
     }
 
-    *iv_length = operation->iv_size;
-    status = psa_cipher_set_iv( operation, iv, *iv_length );
+    status = psa_driver_wrapper_cipher_generate_iv( operation,
```

Comment:
I thought the convention was to let the PSA core draw random bytes, and then just call the `set_iv` wrapper? I.e. get rid of the `generate_iv` driver wrapper call, since it doesn't make sense to have this be the responsibility of the drivers. At least that's what I was told half a year ago by @gilles-peskine-arm .

Reply:
Indeed, I don't think secure elements insist on generating their own IV, so the core might as well do it: it simplifies the driver interface.